File: solutions/0236-lowest-common-ancestor-of-a-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:

    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(!root or root==p or root==q){
            return root;
        }
        TreeNode* rightAns = lowestCommonAncestor(root->right,p,q);
        TreeNode* leftAns = lowestCommonAncestor(root->left,p,q);
        if(rightAns and leftAns){
            return root;
        }
        if(rightAns) return rightAns;
        if(leftAns) return leftAns;
        return nullptr;
    }
};

File: solutions/0236-lowest-common-ancestor-of-a-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root is q or root is p or not root: 
            return root
        a = self.lowestCommonAncestor(root.left,p,q)
        b = self.lowestCommonAncestor(root.right,p,q)
        if a and b:
            return root
        if a:
            return a
        return b
        
